Attack system

SamuelllG opened the pull request:
- implemented reusable attack system for player and enemies
- adapted playercontroller and enemycontroller to use attacklogic
- implemented IDamageable and IMovable interface (IMovable could also be an abstract class and inherit 4-5 methods which are the same for player and enemy)
- implemented hit marker / animation replacement as material (color) change (this is in Stats class which will be replaced with real stats class)

SamuelllG committed:
implemented attack system for the player

SamuelllG committed:
updated player prefab

SamuelllG committed:
Merge branch "npcs_movement" into this branch

SamuelllG committed:
extracted attack system

SamuelllG committed:
adapted EnemyController and added AttackLogic to enemies, updated prefabs, implemented temp hit marker / hit animation replacement

SamuelllG committed:
created Interfaces namespace and sorted scripts

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+    private void Awake()
+    {
+        // init components
+        _movable = GetComponent<IMovable>();
```
Always a good idea to add `[RequiresComponent(typeOf(IMovable))` to the class, if you use GetComponent in the Awake

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+        _movable = GetComponent<IMovable>();
+
+        // -----temp replacement for animation-----
+            _gameObjectRenderer = gameObject.GetComponentInChildren<MeshRenderer>();
```
Please use a reference:(

SamuelllG replied:
over inspector? but it's just temporary...

Kaes3kuch3n replied:
For temporary stuff I'd suggest adding a //TODO or //BUG so that we can find stuff like this easily later

DavidPgl replied:
It only takes a minute. Don't be lazy:P

SamuelllG replied:
haha you're talking about two different things ^^
I made it serializable, Luis talks about the whole animation / temporary color thingy

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+                if (success == true)
+                {
+                    // deal damage
+                    IDamageable damageable = (IDamageable)_target.GetComponent<IDamageable>();
```
You could define _target as IDamageable and save one GetComponent call. GetComponent is very heavy. If you can avoid it, try to.

SamuelllG replied:
How do I save one GetComponent call then? the only point where I call GetComponent is when the hit was performed successfully, if I define _target as IDamageable, I'd have to GetComponent everytime independent of the hit being successful or not. Also this would cause trouble at different places like calculating the distance etc.

DavidPgl replied:
_target is of type GameObject and will be set at some point. Is there a reason why you can't define it as a IDamageable from the beginning? Attacking will happen more often than setting the _target, right?

SamuelllG replied:
I don't know if setting the target (clicking on the enemy) will happen less often than a successful hit... especially because I know people that rape the mouse button on attack :D

(I need the gameobject so that I have the position of the enemy and that I can calculate the distance.)

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+    /// Waits (attackTime) for hit to be finished and returns the result.
+    /// </summary>
+    /// <returns>Null if hit not finished, true if hit successful performed, false if unsuccessful</returns>
+    private bool? PerformHit()
```
Probably personal preference but not the biggest fan of making bool to a three-state type. This is probably one of the few cases where using it is still readable. Defining three Enums for this would probably look nicer though:
`AttackStatus.Hit`
`AttackStatus.Miss`
`AttackStatus.NotFinished`
or something like that.
Also preferences but Unity offers **Coroutines**, in which you can wait for a time or a specified event to be completed. But I'm not sure if you can use it here

SamuelllG replied:
Done, but doesn't improve readability in my opinion. Also removed PerformingHit variable and put it all in one: Status.

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+            // animation finished
+            _timer = 0;
+            // target still in range?
+            float distanceToTarget = Vector3.Distance(_target.transform.position, transform.position);
```
You calculate the distance each frame anywhere (line 78), so saving it in a variable would save you one calculation

SamuelllG replied:
Indeed, however I always try to keep the number of instance variables low (matter of readability), and a parameter doesn't really make sense. But I guess, if it's better to make one, I'll make one ^^

DavidPgl commented on Assets/Scripts/Interfaces/IDamageable.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Interfaces
+{
+    public interface IDamageable
```
Would an abstract class be useful here? I mean enemies and the player both share health and the "loosing health behavior".

SamuelllG replied:
I thought about that as well, maybe we have to discuss this. In my opinion not a matter of this PR, we can create a separate issue for this :)

DavidPgl commented on Assets/Scripts/Interfaces/IMovable.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Interfaces
+{
+    public interface IMovable
```
Same as above. Only if the methods do the same thing of course.

SamuelllG replied:
yep

DavidPgl commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -83,30 +102,106 @@ private void Update()
-                // Create click point effect
-                Instantiate(clickEffect, hit.point + Vector3.up * 5, Quaternion.identity);
+                GameObject objectHit = hit.collider.gameObject;
+                IDamageable damageable = (IDamageable)objectHit.GetComponent<IDamageable>();
```
Is the cast really necessary?

SamuelllG replied:
nope xD

DavidPgl commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -83,30 +102,106 @@ private void Update()
+            return FaceTarget(target, shouldTurn, out f);
+        }
+
+        // ----- Note: same in EnemyController --> make IMovable abstract class and inherit? ------
```
Sounds good to me

SamuelllG replied:
we should a separate issue for this

DavidPgl commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -83,30 +102,106 @@ private void Update()
+        public bool FaceTarget(GameObject target, bool shouldTurn)
+        {
+            float f;
+            return FaceTarget(target, shouldTurn, out f);
```
```suggestion
            return FaceTarget(target, shouldTurn, out _);
```
Then you can delete the local variable

DavidPgl commented on Assets/Scripts/Player/Stats.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
Namespace missing

SamuelllG replied:
yep, I will integrate this into PlayerState

DavidPgl commented on Assets/Scripts/Player/Stats.cs:
```diff
@@ -0,0 +1,64 @@
+    /// </summary>
+    private void Update()
+    {
+        if (_hit)
```
```suggestion
        if (!_hit) return
```
This way it will skip the rest

SamuelllG replied:
and the advantage is? skip one if statement? I rather take the readability, thanks 😄 (I changed it a bit)

DavidPgl commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,175 @@
+    /// Starts the attack.
+    /// </summary>
+    /// <param name="target">The target to be attacked</param>
+    public void StartAttack(GameObject target)
```
```suggestion
    public void StartAttack(IDamageable target)
```
Make it as specific as possible

SamuelllG replied:
as stated above, same thing

Kaes3kuch3n replied:
Could you please elaborate on that? It's not clear at all to me what you mean by that/what you're referring to 😅 

Reply:
You sound like Prof. WeWu :D
Since I couldn't reply, here with an edit: I added a few words to the summary. "If the target is not in attacking range, chases it. If the target is near performs a hit on it and either resets or continues attacking." -- just the normal "attack cycle".

SamuelllG replied:
Oh okay it also didn't show me that you commented on existing comments... what I'm talking about is the discussion of whether to use IDamageable as target or GameObject (see the only open discussion above)

SamuelllG committed:
fixed PR and refactored AttackLogic

SamuelllG committed:
Merge branch 'master' into attack_system

SamuelllG committed:
Merge additions

SamuelllG committed:
fix new AttackStatus

SamuelllG commented:
I also merged master which was a bit of a pain 😅 and fixed everything. I added the abstract class discussion as a task as well as implementing trigger colliders for the NPCs :)

Kaes3kuch3n commented on Assets/Scripts/AttackLogic.cs:
```diff
@@ -0,0 +1,196 @@
+        if (Status == AttackStatus.Hit)
+        {
+            // deal damage
+            IDamageable damageable = (IDamageable)_target.GetComponent<IDamageable>();
```
I'm pretty sure this cast ist redundant, since GetComponent<IDamageable> returns an object of type IDamageable 🤔

SamuelllG replied:
for sure, ye I forgot that here

Kaes3kuch3n commented on Assets/Scripts/EnemyController.cs:
```diff
@@ -150,27 +164,65 @@ private void MoveToNextPoint()
+    public bool FaceTarget(GameObject target, bool shouldTurn)
+    {
+        float f;
+        return FaceTarget(target, shouldTurn, out f);
```
If you don't wanna use an out variable, you can simply use `out _` to discard the result

SamuelllG replied:
forgot that here as well

Kaes3kuch3n commented on Assets/Scripts/Interfaces/IMovable.cs:
```diff
@@ -0,0 +1,17 @@
+
+        bool FaceTarget(GameObject target, bool shouldTurn, out float difference);
+
+        bool FaceTarget(GameObject target, bool shouldTurn);
```
Feels kinda redundant with the other FaceTarget method defined right above it, since we can just discard the out value if we don't need it

SamuelllG replied:
good point

Kaes3kuch3n commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -114,26 +132,101 @@ private void SetUpControls()
-            // Get ground position from mouse click
-            if (Physics.Raycast(clickRay, out RaycastHit hit, 10000, moveClickLayers))
+            // only change target / move, if not performing a hit
+            if (Physics.Raycast(clickRay, out RaycastHit hit, 10000) && _attackLogic.Status == AttackLogic.AttackStatus.None)
```
Would maybe change this in the future to make it possible to cancel attacks by moving away from the enemy. Doesn't really feel like the player is in control like this 😬

SamuelllG replied:
We can discuss about that. It might be useful, idk. It's a tiny change howsoever.

Kaes3kuch3n commented on Assets/Scripts/Player/PlayerController.cs:
```diff
@@ -154,11 +247,11 @@ private void RotateCamera(InputAction.CallbackContext obj)
-            _cameraPosition = new Vector3((float) Math.Sin(radian) * cameraDistance.x, cameraDistance.y,
-                (float) -Math.Cos(radian) * cameraDistance.x);
+            _cameraPosition = new Vector3((float)Math.Sin(radian) * cameraDistance.x, cameraDistance.y,
+                (float)-Math.Cos(radian) * cameraDistance.x);
```
Why did you remove the spaces? 😂

SamuelllG replied:
I didn't, must have come from merging.

Kaes3kuch3n commented on Assets/Scripts/PlayerManager.cs:
```diff
@@ -0,0 +1,34 @@
+
+public class PlayerManager : MonoBehaviour
+{
+    #region Singleton
```
Could you switch to David's Singleton base class for this logic?

SamuelllG replied:
didn't see there is already one - switched

SamuelllG committed:
fix PR

DavidPgl commented on Assets/Scripts/EnemyController.cs:
```diff
@@ -47,9 +52,13 @@ public class EnemyController : MonoBehaviour
+    // component references
     private NavMeshAgent _agent;
     private WanderAI _wanderAI;
+    private AttackLogic _attackLogic;
```
So this script accesses AttackLogic and AttackLogic accesses the EnemyController (with IMovable). Does the AttackLogic really care about IMovable? Wouldn't it be enough if the EnemyController activates the attacking and stops moving?

SamuelllG replied:
That's a good point, but AttackLogic needs IMovable anyway, so that it has a target which it can turn too and which it can check the position after completing the hit :)

I look at it like this: 
You click on the ground --> you move --> Controller
You click on an enemy --> you attack --> AttackLogic
If you have to move in order to attack the enemy that's part of the AttackLogic. Otherwise you'd have to implement something like "target out of range" when clicking on an enemy far away.

DavidPgl replied:
Ahh I see, I was always thinking this way:
You click on something: ground or enemy or resource (Not in range) -->Move to it --> Movable
You click on something (In range) --> Trigger action (e.g. attack or collect) --> specific script/component
Both works I guess

DavidPgl commented on Assets/Scripts/Managers/PlayerManager.cs:
```diff
@@ -1,21 +1,16 @@
     public class PlayerManager : Singleton<PlayerManager>
     {
-        [SerializeField] private GameObject player;
+        [SerializeField] private GameObject _player;
```
SerializedFields are normaly without underscore. Was already correct:)

DavidPgl replied:
```suggestion
        [SerializeField] private GameObject player;
```

SamuelllG replied:
oh ye sure :joy: my bad

SamuelllG replied:
for a singleton I don't expect a serialize

DavidPgl commented on Assets/Scripts/Player/EnemyState.cs:
```diff
@@ -0,0 +1,63 @@
+
+    [Tooltip("The MeshRenderer of the graphics object of the player")]
+    [SerializeField]
+    private MeshRenderer _gameObjectRenderer;
```
Stick with SerializedField without underscore

DavidPgl replied:
```suggestion
    private MeshRenderer gameObjectRenderer;
```

SamuelllG replied:
dang, what did I do

DavidPgl commented on Assets/Scripts/Player/PlayerState.cs:
```diff
@@ -21,7 +21,24 @@ public class PlayerState : MonoBehaviour, IDamageable
+
+        [Tooltip("The MeshRenderer of the graphics object of the player")]
+        [SerializeField]
+        private MeshRenderer _gameObjectRenderer;
```
```suggestion
        private MeshRenderer gameObjectRenderer;
```

DavidPgl commented:
Das ist das letzte :P

SamuelllG committed:
fix PR
